src/main/java/org/opensearch/ad/cluster/ADClusterEventListener.java

```diff
@@ -128,30 +128,26 @@ public void clusterChanged(ClusterChangedEvent event) {
 
             if (dataNodeAdded || dataNodeRemoved) {
                 hashRing.addNodeChangeEvent();
-                boolean finalDataNodeAdded = dataNodeAdded;
                 hashRing.buildCircles(delta, ActionListener.wrap(hasRingBuildDone -> {
                     if (hasRingBuildDone) {
                         LOG.info("Build AD version hash ring successfully");
-                        if (finalDataNodeAdded) {
-                            // Once hash ring rebuilt done, the removed node is not in hash ring, so we just stop model
-                            // for data node added case.
-                            String localNodeId = event.state().nodes().getLocalNode().getId();
-                            Set<String> modelIds = modelManager.getAllModelIds();
-                            for (String modelId : modelIds) {
-                                Optional<DiscoveryNode> node = hashRing.getOwningNodeWithSameLocalAdVersionForRealtimeAD(modelId);
-                                if (node.isPresent() && !node.get().getId().equals(localNodeId)) {
-                                    LOG.info(REMOVE_MODEL_MSG + " {}", modelId);
-                                    modelManager
-                                        .stopModel(
-                                            modelManager.getDetectorIdForModelId(modelId),
-                                            modelId,
-                                            ActionListener
-                                                .wrap(
-                                                    r -> LOG.info("Stopped model [{}] with response [{}]", modelId, r),
-                                                    e -> LOG.error("Fail to stop model " + modelId, e)
-                                                )
-                                        );
-                                }
+                        String localNodeId = event.state().nodes().getLocalNode().getId();
+                        Set<String> modelIds = modelManager.getAllModelIds();
+                        for (String modelId : modelIds) {
```

Comment:
Previously, we do it whenever hash ring rebuild successfully (depending on cool down is past and other concurrent conditions).  Now, you would build the hash ring without the cooldown.  So hasRingBuildDone only means historical analysis's hash ring change, but real time does not necessarily so.  So real time analysis might be use the old hash ring while we clear models here.  Would you consider moving the logic to HashRing when the real time hash ring actually changes?

Reply:
The cluster node added/removed event is limited, generally it only happens when cluster starts or scales. It may have some extra checks during the cool down period, but that impact should be minor. How about add some TODO and fix it later?